### PR TITLE
Dataflow pipeline

### PR DESCRIPTION
**Pull request by fprost**
Creating the structure for the Dataflow script.

Our goal here is to add to estimator/ a template for a Dataflow job. The pipeline will be simple (just train/eval split) but it will be useful both for teaching purposes and for using it as a template for all Dataflow projects (typical structure).

### CODE REVIEW

**Comment by fprost** [comment]
@cfezequiel @dan-anghel

File: census/estimator/preprocessing/preprocess.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+import csv
+
+import apache_beam as beam
+import tensorflow as tf
+
+
+class CsvFileSource(beam.io.filebasedsource.FileBasedSource):
+
+  def read_records(self, file_name, range_tracker):
```
Comment (cfezequiel) [inline census/estimator/preprocessing/preprocess.py]: `range_tracker` appears unused. You can rename it to `unused_range_tracker` if it's needed in the signature, or write the following in the function body:
```
_ = range_tracker
```

File: census/estimator/run_preprocessing.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+from preprocessing import preprocess
+
+
+def _parse_arguments(argv):
+  """Parses command line arguments."""
+  parser = argparse.ArgumentParser(
+      description='Runs preprocessing on census train data.')
+  parser.add_argument(
+      '--project_name',
```
Comment (cfezequiel) [inline census/estimator/run_preprocessing.py]: Minor nit: consider using `project_id` as it seems to be the common term used to refer to the Cloud project in documentation.

**Review by cfezequiel — APPROVED** [review]

**Comment by fprost** [comment]
@gogasca @dizcology @thedriftofwords @happyhuman

Initial code review for the feature request [here](https://github.com/GoogleCloudPlatform/cloudml-samples/issues/315). We will write code in this fork, make small PRs here and merge with master when ready. 

Thank you

File: census/estimator/requirements.txt [hunk 3]
```diff
@@ -0,0 +1,4 @@
+google-cloud-dataflow
+apache-beam[gcp]
+tensorflow==1.10.0
+tensorboard==1.9.0
```
Comment (gogasca) [inline census/estimator/requirements.txt]: Any reason why we use TB 1.9 vs 1.10 ?

File: census/estimator/requirements.txt [hunk 4]
```diff
@@ -0,0 +1,4 @@
+google-cloud-dataflow
+apache-beam[gcp]
+tensorflow==1.10.0
```
Comment (gogasca) [inline census/estimator/requirements.txt:3]: New runtime 1.12 will be available this week.

File: census/estimator/run_preprocessing.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+  options = {'project': ml_project}
+
+  if args.cloud:
+    if not args.job_name:
+      raise ValueError('Job name must be specified for cloud runs.')
+    options.update({
+        'job_name': args.job_name,
+        'num_workers': int(config.get('num_workers')),
+        'max_num_workers': int(config.get('max_num_workers')),
+        # 'setup_file': os.path.abspath(os.path.join(
```
Comment (gogasca) [inline census/estimator/run_preprocessing.py]: clean up this?

File: census/estimator/README.md [hunk 6]
```diff
@@ -0,0 +1,5 @@
+# TODO(fprost): Write README
+
+Dataflow:
+The current code uses the test set to control overfitting or to do hyper parameter tuning, while we typically recommend to use an evaluation set for this.
```
Comment (gogasca) [inline census/estimator/README.md]: nit: hyper parameter -> hyperparameter

File: census/estimator/bin/run.preprocess.sh [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Convenience script for running preprocessing jobs."""
+
+
+PROJECT_ID=$(gcloud config list --format 'value(core.project)' 2>/dev/null)
+JOB_NAME=preprocessing-${DATE_TIME}-${USER}
+DATAFLOW_DIR=gs://census-example/preprocessing/${JOB_NAME}
```
Comment (gogasca) [inline census/estimator/bin/run.preprocess.sh]: Maybe use a place holder for people to choose their own BUCKET_NAME or an argument?:
```
BUCKETNAME='census-example'
DATAFLOW_DIR=gs://$BUCKETNAME/preprocessing/${JOB_NAME}
```

**Review by gogasca — CHANGES_REQUESTED** [review]

File: census/estimator/requirements.txt (hunk 3 above)
Comment (fprost) [reply census/estimator/requirements.txt]: Modified to 1.10

File: census/estimator/requirements.txt (hunk 4 above)
Comment (fprost) [reply census/estimator/requirements.txt:3]: Good to know. Will update this when available

File: census/estimator/run_preprocessing.py (hunk 5 above)
Comment (fprost) [reply census/estimator/run_preprocessing.py]: done

File: census/estimator/README.md (hunk 6 above)
Comment (fprost) [reply census/estimator/README.md]: done

File: census/estimator/bin/run.preprocess.sh (hunk 7 above)
Comment (fprost) [reply census/estimator/bin/run.preprocess.sh]: done.

**Review by gogasca — APPROVED** [review]
LGTM

**Pull request merged**